[NEW] Disable user typing status (#12423)

computerManiac opened the pull request:
<!-- INSTRUCTION: Your Pull Request name should start with one of the following tags -->
<!-- [NEW] For new features -->
<!-- [FIX] For bug fixes -->
<!-- [BREAK] For pull requests including breaking changes -->

<!-- INSTRUCTION: Inform the issue number that this PR closes, or remove the line below -->
Closes #12423 

<!-- INSTRUCTION: Link to a https://github.com/RocketChat/docs PR with added/updated documentation or an update to the missing/outdated documentation list, see https://rocket.chat/docs/contributing/documentation/  -->

<!-- INSTRUCTION: Tell us more about your PR with screen shots if you can -->
Rocket.Chat by default shows the **_user is typing_** status whenever a user starts typing as seen below.

![screenshot_2018-10-26 dev-server](https://user-images.githubusercontent.com/24594035/47556063-1097b900-d92b-11e8-8c80-ad78a695a554.png)

This fix allows users to disable it. To disable it, go to _packages/rocketchat-ui-message/client/messageBox.js_ and on line 7 we have the following constant defined:
```javascript
const disableUserTyping = false;
```
Changing the value to **true** will disable user typing status.

computerManiac commented:
> However, I do not think this should be something that can be turned off. Realizing that someone else is typing is very vital to several conversation flows and thus not having it would actually be a hindrance instead of a help.

I think we should let users have the ability to disable this, but if an admin feels this feature is vital for conversation then he can deny users from disabling it on the server.

graywolf336 commented:
I'm going to stand behind my original statement and say that I'm not sure we should even consider this feature. Otherwise, it can inadvertently lead to miscommunications.

computerManiac commented:
Should I close the PR ?

graywolf336 commented:
No sir, we will await responses from the rest of the Rocket.Chat team 🙂

reetp commented:
Personally I can't see the harm in this if the ultimate arbiter is the admin, and someone has already written the code.

We are all different and have differing requirements.

computerManiac commented:
I have added the option under _Message_ section of Admin Settings to Allow Users to disable typing status. The users can disable it under account settings, if allowed by the admin otherwise the _Disable User Typing_ remains false. There is however a slight problem which I am unable to fix. I have added the corresponding entries in _en.i18n.json_ and the text displayed correctly for the first run but for subsequent runs, the text didn't display correctly. Disabling user typing status unsubscribes the user from the typing stream, so he/she won't be able to see if other users are typing.

vynmera commented:
why not an administration key? in Message or w/e?

reetp commented:
Any more comments on this from the devs?

tassoevan commented on packages/rocketchat-ui/client/lib/msgTyping.js:
```diff
@@ -24,6 +31,10 @@ export const MsgTyping = (function() {
 
 	const addStream = function(room) {
+		if (disableUserTyping === true) {
+			RocketChat.Notifications.unRoom(room, 'typing', function() { return; });
```
The third argument must be the original event listener callback attached to the event, e.g.,

```js
// The event listener callback (cb) is attached to the event
const cb = () => { ... };
RocketChat.Notifications.onRoom(room, 'typing', cb);

...

// The same event listener callback is detached
RocketChat.Notifications.unRoom(room, 'typing', cb);
```

tassoevan commented on packages/rocketchat-ui/client/lib/msgTyping.js:
```diff
@@ -12,6 +12,13 @@ export const MsgTyping = (function() {
+		if (!Meteor.userId()) {
+			return;
+		}
+		disableUserTyping = RocketChat.getUserPreference(Meteor.userId(), 'disableUserTyping');
```
If `Allow_DisableUserTyping` changes from `true` to `false`, a user can still have the `disableUserTyping` preference set to `true`.

```suggestion
		disableUserTyping = RocketChat.settings.get('Allow_DisableUserTyping') === true 
 ? RocketChat.getUserPreference(Meteor.userId(), 'disableUserTyping') : false;
```

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/RocketChat/Rocket.Chat?pullRequest=12431) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/RocketChat/Rocket.Chat?pullRequest=12431) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: tassoevan<br/>:x: computerManiac<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/RocketChat/Rocket.Chat?pullRequest=12431) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/RocketChat/Rocket.Chat?pullRequest=12431) <br/>All committers have signed the CLA.

rodrigok commented:
Closing this one in favor of https://github.com/RocketChat/Rocket.Chat/pull/13232

Thanks
